### DISCUSSION TITLE
train/val from csv files

### DISCUSSION THREAD

**Discussion by ahmed-moubtahij** [explosion/spaCy#8277] [Help: Coding & Implementations] [answered]
My training data consists of a simple column of sentences in the form of a csv file. `corpora.train` apparently expects the path to a `.spacy` binary which afaiu is obtainable with the `convert` command. My problem is that this command apparently only handles formats from this list: https://spacy.io/api/cli#converters but not `.csv` extensions.

To accomodate one of the types eligible for conversion, say `.json` I'm thinking of adding an "ID" column to my csv with just numbers and create a `.json`  file from it with IDs as keys and sentences as values. Is this the expected approach?

**Comment by adrianeboyd** [comment, marked as answer]
The [spaCy v2 JSON format](https://v2.spacy.io/api/annotation#json-input) is very specific JSON format and is probably not what you want to use.

If your data is not in one of the formats supported by `spacy convert`, a better place to start might be a demo project for the type of component you want to train. For textcat, e.g.:

https://github.com/explosion/projects/tree/v3/pipelines/textcat_demo

The demo convert script shows how to create `Doc` objects from a simple JSONL format (a script for CSV would be very similar) and save those in a `DocBin` as a `.spacy` file:

https://github.com/explosion/projects/blob/v3/pipelines/textcat_demo/scripts/convert.py

You can browse other demos here:

https://github.com/explosion/projects/tree/v3/pipelines/

**Reply by ahmed-moubtahij** [reply]
Thank you for the reply.

https://github.com/explosion/projects/blob/b9db3c3883f7403c5e91b7cfc421ec7dde4a8b37/pipelines/textcat_demo/scripts/convert.py#L14 uses `srsly.read_jsonl` and from https://github.com/explosion/srsly:

> Currently supports JSON, JSONL, MessagePack, Pickle and YAML.

Whereas my input file is a csv and afaict so there is no `read_csv` equivalent in `srsly` 😕

**Reply by ahmed-moubtahij** [reply]
@svlandeg Thank you for marking this as the answer, but the proposed code example is based on `srsly` which does not handle `csv` formats...

**Reply by svlandeg** [reply]
I think this is the key point in Adriane's answer: 
> The demo convert script shows how to create `Doc` objects ... and save those in a `DocBin` as a `.spacy` file

How you parse the data out of your custom CSV files is really up to you, and there are plenty of Python libraries to read in CSV (and examples on StackOverflow). Once you've read & parsed your data, you can set the correct field on the `Doc` objects, and then you can save the `.spacy` files with `DocBin` as shown in the example.

**Reply by ahmed-moubtahij** [reply]
Right, I had gotten that that part was key but I had also wrongly assumed that the serialization from `srsly` was necessary; it wasn't, I could just parse out a string to pass directly to `nlp.make_doc()`. I have my `.spacy` binaries now, thank you :)

**Reply by svlandeg** [reply]
Perfect, happy to hear it!
